File: 5. Smallest Multiple/Ruby/solution.rb
# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

# What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?


def lowest_common_multiple(n)
  distinct_primes = []
  prime_numbers = (2..n).select{|x| is_prime(x)}
  (2..20).each do |x|
    prime_numbers.each |y|
      if x % y == 0
        
      end
    end
  end
end

def all_factors(n)
  factors = []
  (2..Math.sqrt(n)).each do |x|
    if n % x == 0
      factors << x 
      factors << n/x
    end
  end
  factors 
end

def is_prime(n)
  all_factors(n).empty?
end

File: 5. Smallest Multiple/Javascript/solution.js
# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

# What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

var lowestCommonMultiple = function(n){
  var primeNumbers = [];
  var distinctPrimes;
  for(var i = 2; i < n; i++){
    if(isPrime(i)){
      primeNumbers.push(i)
    }
  }
  distinctPrimes = primeNumbers.map(function(x){
    var exponent = 1;
    var prime = x;
    while(prime < n){
      if(Math.pow(x,exponent) > prime) {
        prime = Math.pow(x,exponent)
      }
      exponent += 1
    }
    return prime / x;
  })
  return distinctPrimes.reduce(function(x,y){return x * y;})
};

var allFactors = function(n){
  var factors = [];
  for(var i = 2; i <= Math.sqrt(n); i++){
    if(n % i === 0){
      factors.push(i)
      factors.push(n / i)
    }
  }
  return factors;
};

var isPrime = function(n){
  return (allFactors(n).length === 0) ? true : false;
};